src/DomainServices/ProjectService.ts

```diff
@@ -0,0 +1,255 @@
+    await this.validateManuscriptIDs(projectID, models)
+
+    await this.containerRepository.bulkUpsert(models)
+  }
+
+  public async getProjectModels(projectID: string): Promise<Model[]> {
+    return await this.containerRepository.getContainerResources(projectID, null)
+  }
+
+  public async updateUserRole(
+    projectID: string,
+    userID: string,
+    role: ProjectUserRole
+  ): Promise<void> {
+    const project = await this.getProject(projectID)
+    const user = await this.userRepository.getOne({ connectUserID: userID })
+
+    if (!user) {
+      throw new ValidationError('Invalid user id', user)
+    }
+
+    if (user._id === '*' && (role === 'Owner' || role === 'Writer')) {
+      throw new ValidationError('User can not be owner or writer', user._id)
+    }
+
+    if (ProjectService.isOnlyOwner(project, user._id)) {
+      throw new UserRoleError('User is the only owner', role)
+    }
+
+    const updated = {
+      _id: projectID,
+      owners: project.owners.filter((u) => u !== userID),
+      writers: project.writers.filter((u) => u !== userID),
+      viewers: project.viewers.filter((u) => u !== userID),
+      editors: project.editors ? project.editors.filter((u) => u !== userID) : [],
+      annotators: project.annotators ? project.annotators.filter((u) => u !== userID) : [],
+    }
+
+    switch (role) {
+      case ProjectUserRole.Owner:
+        updated.owners.push(userID)
+        break
+      case ProjectUserRole.Writer:
+        updated.writers.push(userID)
+        break
+      case ProjectUserRole.Viewer:
+        updated.viewers.push(userID)
+        break
+      case ProjectUserRole.Editor:
+        updated.editors.push(userID)
+        break
+      case ProjectUserRole.Annotator:
+        updated.annotators.push(userID)
+        break
+    }
+
+    await this.containerRepository.patch(projectID, updated)
+  }
+
+  public async generateAccessToken(
+    projectID: string,
+    userID: string,
+    scope: string
+  ): Promise<string> {
+    const scopeInfo = config.scopes.find((s) => s.name === scope)
+
+    if (!scopeInfo) {
+      throw new InvalidScopeNameError(scope)
+    }
+
+    const payload = {
+      iss: config.API.hostname,
+      sub: userID,
+      containerID: projectID,
+      aud: scopeInfo.name,
+    }
+
+    const algorithm: Algorithm = scopeInfo.publicKeyPEM === null ? 'HS256' : 'RS256'
+
+    const options = {
+      header: {
+        kid: scopeInfo.identifier,
+        alg: algorithm,
+      },
+      expiresIn: `${scopeInfo.expiry}m`,
+    }
+
+    return jwt.sign(payload, scopeInfo.secret, options)
+  }
+
+  public static isOnlyOwner(project: Project, userID: string): boolean {
+    return project.owners.length === 1 && project.owners[0] === userID
+  }
+
+  private validateContainerIDs(projectID: string, models: any[]) {
+    if (!models.every((m) => m.containerID !== projectID)) {
+      throw new ValidationError(`problem with containerID`, models)
+    }
+  }
```

Comment:
Hello @mnatsheh 
while unit testing, I came across this
shouldn't this be `models: Model[]` instead of `models: any[]` and `_id` instead of `containerID`, because no way models here will have the  property `containerID`. the only call to this function is from `updateProject` which has `models: Model`, so this function will never throw an error

Reply:
another question is, shouldn't this be 
```  if (!models.every((m) => m.containerID === projectID) ```
instead of 
``` if (!models.every((m) => m.containerID !== projectID)```
shouldn't all models belong to the same container?

Reply:
The problem is that `Model` doesn't have the `containerID` property, so to make things simple I used `any[]`. If there is an alternative feel free to change it. I don't think checking for `_id` instead of `containerID` is right since these models are not only the project object. I'm not sure if a `Project` object is also there, but if it is then we will need to add an `_id` check _in addition_ to the `containerID` check (e.g. `m.containerID === projectID || m._id === projectID`).

As for the `!==`, I think you're right so please change it.